File: tests/lua/dbus/test_error.lua
#!/usr/bin/env lua
--[[
Copyright (c) 2026 Huawei Technologies Co., Ltd.
openUBMC is licensed under Mulan PSL v2.
You can use this software according to the terms and conditions of the Mulan PSL v2.
You may obtain a copy of Mulan PSL v2 at:
        http://license.coscl.org.cn/MulanPSL2
THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
See the Mulan PSL v2 for more details.
]]

local lu = require('luaunit')
local ldbus = require('ldbus')

-- 测试 ldbus 错误处理
TestError = {}

function TestError:setUp()
    self.blocking = ldbus.blocking
    self.message = ldbus.message
end

function TestError:tearDown()
    if self.conn then
        pcall(function() self.conn:close() end)
        self.conn = nil
    end
end

function TestError:test_01_call_without_connection()
    self.conn = self.blocking.open_user()
    -- 不调用 start()，但 open_user 会自动建立连接
    -- 这个测试需要验证未完全初始化的连接行为
    
    -- 注意：blocking.open_user() 实际上会建立底层连接
    -- 所以这个测试的预期可能需要调整
    local inner_conn = self.conn.conn
    lu.assertNotNil(inner_conn, "conn should be created")
    
    -- 尝试调用，可能成功或失败取决于实现
    local success, err = pcall(function()
        local msg = self.message.new_method_call(
            "org.freedesktop.DBus",
            "/org/freedesktop/DBus",
            "org.freedesktop.DBus",
            "ListNames"
        )
        inner_conn:send_with_reply_and_block(msg, 1000)
    end)
    
    -- 只验证调用有返回结果，不强制要求失败
    lu.assertNotNil(success, "Should have a result")
end

function TestError:test_02_call_invalid_service()
    self.conn = self.blocking.open_user()
    
    local success, err = pcall(function()
        local inner_conn = self.conn.conn
        local msg = self.message.new_method_call(
            "org.invalid.Service",
            "/",
            "org.invalid.Interface",
            "Method"
        )
        inner_conn:send_with_reply_and_block(msg, 1000)
    end)
    
    lu.assertEquals(success, false, "Calling invalid service should fail")
    lu.assertNotNil(err, "Should return error message")
end

function TestError:test_03_invalid_timeout_parameter()
    self.conn = self.blocking.open_user()
    
    local success, err = pcall(function()
        local inner_conn = self.conn.conn
        local msg = self.message.new_method_call(
            "org.freedesktop.DBus",
            "/org/freedesktop/DBus",
            "org.freedesktop.DBus",
            "ListNames"
        )
        -- 传入字符串作为 timeout（应该失败）
        inner_conn:send_with_reply_and_block(msg, "invalid")
    end)
    
    lu.assertEquals(success, false, "Invalid timeout type should fail")
end

function TestError:test_04_missing_required_parameters()
    local success, err = pcall(function()
        -- 尝试创建不完整的 message
        self.message.new_method_call("org.freedesktop.DBus")
    end)
    
    lu.assertEquals(success, false, "Missing parameters should fail")
end

function TestError:test_05_short_timeout()
    self.conn = self.blocking.open_user()
    
    -- 使用很短的超时时间
    local start_time = os.clock()
    local success, err = pcall(function()
        local inner_conn = self.conn.conn
        local msg = self.message.new_method_call(
            "org.freedesktop.DBus",
            "/org/freedesktop/DBus",
            "org.freedesktop.DBus",
            "ListNames"
        )
        inner_conn:send_with_reply_and_block(msg, 1)  -- 1ms 超时
    end)
    local elapsed = os.clock() - start_time
    
    -- 可能成功（如果响应很快）或超时
    lu.assertNotNil(success, "Timeout test completed")
end

function TestError:test_06_connection_lifecycle()
    self.conn = self.blocking.open_user()
    
    -- 正常调用
    local inner_conn = self.conn.conn
    local msg1 = self.message.new_method_call(
        "org.freedesktop.DBus",
        "/org/freedesktop/DBus",
        "org.freedesktop.DBus",
        "ListNames"
    )
    local reply1 = inner_conn:send_with_reply_and_block(msg1)
    lu.assertNotNil(reply1, "Call before close should work")
    
    -- 关闭连接
    self.conn:close()
    
    -- 尝试在关闭后调用
    local success, err = pcall(function()
        local msg2 = self.message.new_method_call(
            "org.freedesktop.DBus",
            "/org/freedesktop/DBus",
            "org.freedesktop.DBus",
            "ListNames"
        )
        inner_conn:send_with_reply_and_block(msg2)
    end)
    
    lu.assertEquals(success, false, "Call after close should fail")
end

function TestError:test_07_error_message_content()
    self.conn = self.blocking.open_user()
    
    -- 测试调用无效服务来获取错误消息
    local success, err = pcall(function()
        local inner_conn = self.conn.conn
        local msg = self.message.new_method_call(
            "org.invalid.NonExistentService",
            "/",
            "org.invalid.Interface",
            "Method"
        )
        inner_conn:send_with_reply_and_block(msg, 100)
    end)
    
    lu.assertEquals(success, false, "Calling invalid service should fail")
    lu.assertEquals(type(err), "string", "Error should be a string")
    lu.assertIsTrue(string.len(err) > 0, "Error message should not be empty")
end

-- 只有当直接运行此文件时才执行测试
if arg and arg[0]:match("test_error%.lua$") then
    os.exit(lu.LuaUnit.run())
end

return TestError

File: tests/dbus/test_error.cpp
/*
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * openUBMC is licensed under Mulan PSL v2.
 * You can use this software according to the terms and conditions of the Mulan
 * PSL v2. You may obtain a copy of Mulan PSL v2 at:
 *         http://license.coscl.org.cn/MulanPSL2
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY
 * KIND, EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
 * NON-INFRINGEMENT, MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE. See the
 * Mulan PSL v2 for more details.
 */

#include <mc/dbus/error.h>
#include <mc/dict.h>

#include <gtest/gtest.h>

using namespace mc::dbus;

// 测试默认构造
TEST(ErrorTest, DefaultConstructor)
{
    error err;
    EXPECT_FALSE(err.is_set());
}

// 测试复制构造
TEST(ErrorTest, CopyConstructor)
{
    error err1;
    err1.set_error("org.freedesktop.DBus.Error.Failed", "Test error message");

    error err2(err1);
    EXPECT_TRUE(err2.is_set());
    EXPECT_STREQ(err2.name, err1.name);
    EXPECT_STREQ(err2.message, err1.message);
}

// 测试复制构造 - 源对象未设置错误
TEST(ErrorTest, CopyConstructorUnset)
{
    error err1; // 未设置错误
    error err2(err1);
    EXPECT_FALSE(err2.is_set());
}

// 测试复制赋值
TEST(ErrorTest, CopyAssignment)
{
    error err1;
    err1.set_error("org.freedesktop.DBus.Error.Failed", "Test error message");

    error err2;
    err2 = err1;
    EXPECT_TRUE(err2.is_set());
    EXPECT_STREQ(err2.name, err1.name);
    EXPECT_STREQ(err2.message, err1.message);
}

// 测试复制赋值 - 源对象未设置错误
TEST(ErrorTest, CopyAssignmentUnset)
{
    error err1; // 未设置错误
    error err2;
    err2.set_error("org.freedesktop.DBus.Error.Failed", "Test error message");
    err2 = err1; // 从未设置的对象赋值
    EXPECT_FALSE(err2.is_set());
}

// 测试移动赋值
TEST(ErrorTest, MoveAssignment)
{
    error err1;
    err1.set_error("org.freedesktop.DBus.Error.Failed", "Test error message");

    // 确保目标对象是未设置的（默认构造已初始化）
    error err2; // 默认构造，确保未设置错误
    err2 = std::move(err1);
    EXPECT_TRUE(err2.is_set());
    EXPECT_STREQ(err2.name, "org.freedesktop.DBus.Error.Failed");
    EXPECT_STREQ(err2.message, "Test error message");

    // 移动后原对象应该不再设置错误
    EXPECT_FALSE(err1.is_set());
}

// 测试移动构造
TEST(ErrorTest, MoveConstructor)
{
    error err1;
    err1.set_error("org.freedesktop.DBus.Error.Failed", "Test error message");

    error err2(std::move(err1));
    EXPECT_TRUE(err2.is_set());
    EXPECT_STREQ(err2.name, "org.freedesktop.DBus.Error.Failed");
    EXPECT_STREQ(err2.message, "Test error message");

    // 移动后原对象应该不再设置错误
    EXPECT_FALSE(err1.is_set());
}

// 测试 set_error (字符串视图版本)
TEST(ErrorTest, SetErrorStringView)
{
    error err;
    err.set_error("org.freedesktop.DBus.Error.Failed", "Test error message");

    EXPECT_TRUE(err.is_set());
    EXPECT_STREQ(err.name, "org.freedesktop.DBus.Error.Failed");
    EXPECT_STREQ(err.message, "Test error message");
}

// 测试 set_error (带格式化参数版本)
TEST(ErrorTest, SetErrorWithDict)
{
    error    err;
    mc::dict args;
    args["param"] = "value";
    args["count"] = 42;

    err.set_error("org.freedesktop.DBus.Error.Failed", "Error: ${param} = ${count}", args);

    EXPECT_TRUE(err.is_set());
    EXPECT_STREQ(err.name, "org.freedesktop.DBus.Error.Failed");
    EXPECT_NE(std::string(err.message).find("value"), std::string::npos);
    EXPECT_NE(std::string(err.message).find("42"), std::string::npos);
}

// 测试 set_error (空字典)
TEST(ErrorTest, SetErrorWithEmptyDict)
{
    error    err;
    mc::dict args;

    err.set_error("org.freedesktop.DBus.Error.Failed", "Test error message", args);

    EXPECT_TRUE(err.is_set());
    EXPECT_STREQ(err.name, "org.freedesktop.DBus.Error.Failed");
    EXPECT_STREQ(err.message, "Test error message");
}

// 测试 set_error_const
TEST(ErrorTest, SetErrorConst)
{
    error       err;
    const char* name    = "org.freedesktop.DBus.Error.Failed";
    const char* message = "Test error message";

    err.set_error_const(name, message);

    EXPECT_TRUE(err.is_set());
    EXPECT_STREQ(err.name, name);
    EXPECT_STREQ(err.message, message);
}

// 测试错误名称常量
TEST(ErrorTest, ErrorNames)
{
    EXPECT_EQ(std::string(error_names::failed), "org.freedesktop.DBus.Error.Failed");
    EXPECT_EQ(std::string(error_names::no_memory), "org.freedesktop.DBus.Error.NoMemory");
    EXPECT_EQ(std::string(error_names::service_unknown), "org.freedesktop.DBus.Error.ServiceUnknown");
    EXPECT_EQ(std::string(error_names::name_has_no_owner), "org.freedesktop.DBus.Error.NameHasNoOwner");
    EXPECT_EQ(std::string(error_names::no_reply), "org.freedesktop.DBus.Error.NoReply");
    EXPECT_EQ(std::string(error_names::io_error), "org.freedesktop.DBus.Error.IOError");
    EXPECT_EQ(std::string(error_names::bad_address), "org.freedesktop.DBus.Error.BadAddress");
    EXPECT_EQ(std::string(error_names::not_supported), "org.freedesktop.DBus.Error.NotSupported");
    EXPECT_EQ(std::string(error_names::limits_exceeded), "org.freedesktop.DBus.Error.LimitsExceeded");
    EXPECT_EQ(std::string(error_names::access_denied), "org.freedesktop.DBus.Error.AccessDenied");
    EXPECT_EQ(std::string(error_names::auth_failed), "org.freedesktop.DBus.Error.AuthFailed");
    EXPECT_EQ(std::string(error_names::no_server), "org.freedesktop.DBus.Error.NoServer");
    EXPECT_EQ(std::string(error_names::timeout), "org.freedesktop.DBus.Error.Timeout");
    EXPECT_EQ(std::string(error_names::no_network), "org.freedesktop.DBus.Error.NoNetwork");
    EXPECT_EQ(std::string(error_names::address_in_use), "org.freedesktop.DBus.Error.AddressInUse");
    EXPECT_EQ(std::string(error_names::disconnected), "org.freedesktop.DBus.Error.Disconnected");
    EXPECT_EQ(std::string(error_names::invalid_args), "org.freedesktop.DBus.Error.InvalidArgs");
    EXPECT_EQ(std::string(error_names::file_not_found), "org.freedesktop.DBus.Error.FileNotFound");
    EXPECT_EQ(std::string(error_names::file_exists), "org.freedesktop.DBus.Error.FileExists");
    EXPECT_EQ(std::string(error_names::unknown_method), "org.freedesktop.DBus.Error.UnknownMethod");
    EXPECT_EQ(std::string(error_names::unknown_object), "org.freedesktop.DBus.Error.UnknownObject");
    EXPECT_EQ(std::string(error_names::unknown_interface), "org.freedesktop.DBus.Error.UnknownInterface");
    EXPECT_EQ(std::string(error_names::unknown_property), "org.freedesktop.DBus.Error.UnknownProperty");
    EXPECT_EQ(std::string(error_names::property_read_only), "org.freedesktop.DBus.Error.PropertyReadOnly");
}
